Add support for DNS options

thockin opened the pull request:
Add Support for DNS options (e.g. ndots:3).  This depends on a new libnetwork being vendored in.  I don't know how to do that or if "mere mortals" are allowed to.

@mavenugo 

Regarding #14069

Signed-off-by: Tim Hockin thockin@google.com

thockin committed:
Add support for DNS options

Signed-off-by: Tim Hockin <thockin@google.com>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "14069-dns-options" git@github.com:thockin/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

mavenugo commented:
@thockin :) 
Answer to your question : https://github.com/docker/docker/pull/16038

thockin commented:
great!

On Wed, Sep 2, 2015 at 5:36 PM, Madhu Venugopal <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> :)
> Answer to your question : #16038
> <https://github.com/docker/docker/pull/16038>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16031#issuecomment-137282383>.
>

calavera commented:
@thockin it looks like the signature parser doesn't like the format in this commit. Do you mind to amend with `-s` to let git use a valid format?

thockin commented:
Weird, fixed.  Also rebased.

On Wed, Sep 9, 2015 at 12:42 PM, David Calavera <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> it looks like the signature parser
> doesn't like the format in this commit. Do you mind to amend with -s to
> let git use a valid format?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16031#issuecomment-139024683>.
>

cpuguy83 commented:
Looks pretty good, but would be great if we can do validations on the dns opts input to verify
1) correct format
2) valid options

calavera commented:
My only concern is that we call all option flags with the prefix `-opt`, like `--log-opt` and `--security-opt`. I think we should be consistent and call it `--dns-opt`. Otherwise, I'm onboard with this change.

thockin commented:
@cpuguy83 there's not really a single standard for options strings.  Linux resolv.conf spec defines: ndots:n timeout:n attempts:n rotate
no-check-names inet6 ip6-bytestring ip6-dotint and more.  Some of these are explictly versioned for some glibc versions.  I can not find a hard rule on allowed characters - we could make assumptions, but that makes me nervous.

@calavera will do - look for a push in O(minutes)

thockin commented:
push is up

calavera commented:
nice. The change LGTM.

/cc @mavenugo, @mrjana, @aboch

aboch commented:
The changes in `buildSandboxOptions()` LGTM

albers commented:
bash completion LGTM. Thanks for including it in this PR.
@sdurrheimer, can you please take a look at the zsh completion?

sdurrheimer commented on contrib/completion/zsh/_docker:
```diff
@@ -599,6 +599,7 @@ _docker() {
         "($help)--default-gateway-v6[Container default gateway IPv6 address]:IPv6 address: " \
         "($help)*--dns=-[DNS server to use]:DNS: " \
         "($help)*--dns-search=-[DNS search domains to use]" \
+        "($help)*--dns-opt=-[Set custom DNS options]" \
```
Can you modify this to `"($help)*--dns-opt=-[Set custom DNS options]:DNS option: " \`. This adds a help message to the user.
You can also a modify the `--dns-search` to `"($help)*--dns-search=-[DNS search domains to use]:DNS search: " \` for the same reason.

Also, `--dns-opt` need to be added between line 230 and 231 for the `create` and `run` command.

thockin replied:
Done all around and pushed.

On Fri, Sep 11, 2015 at 12:30 AM, Steve Durrheimer <notifications@github.com
> wrote:

> In contrib/completion/zsh/_docker
> <https://github.com/docker/docker/pull/16031#discussion_r39247116>:
>
> > @@ -599,6 +599,7 @@ _docker() {
> >          "($help)--default-gateway-v6[Container default gateway IPv6 address]:IPv6 address: " \
> >          "($help)*--dns=-[DNS server to use]:DNS: " \
> >          "($help)*--dns-search=-[DNS search domains to use]" \
> > +        "($help)*--dns-opt=-[Set custom DNS options]" \
>
> Can you modify this to "($help)*--dns-opt=-[Set custom DNS options]:DNS
> option: " \. This adds a help message to the user.
> You can also a modify the --dns-search to "($help)*--dns-search=-[DNS
> search domains to use]:DNS search: " \ for the same reason.
>
> Also, --dns-opt need to be added between line 230 and 231 for the create
> and run command.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16031/files#r39247116>.
>

sdurrheimer commented:
Done. I have added a comment.

When this will be merged, I will move the `--dns-opt` to the `daemon` command in PR #15915.

sdurrheimer commented on contrib/completion/zsh/_docker:
```diff
@@ -226,8 +226,9 @@ __docker_subcommand() {
-        "($help)*--dns=-[Set custom dns servers]:dns server: "
-        "($help)*--dns-search=-[Set custom DNS search domains]:dns domains: "
+        "($help)*--dns=-[Set custom DNS servers]:DNS server: "
+        "($help)*--dns-opt=-[Set custom DNS options]:DNS option: " \
```
Typo, need to remove the ` \` (this section of code is a array).

thockin replied:
oops!  Fixed and pushed

On Fri, Sep 11, 2015 at 1:43 AM, Steve Durrheimer <notifications@github.com>
wrote:

> In contrib/completion/zsh/_docker
> <https://github.com/docker/docker/pull/16031#discussion_r39251701>:
>
> > @@ -226,8 +226,9 @@ __docker_subcommand() {
> >          "($help)*--cap-drop=-[Drop Linux capabilities]:capability: "
> >          "($help)--cidfile=-[Write the container ID to the file]:CID file:_files"
> >          "($help)*--device=-[Add a host device to the container]:device:_files"
> > -        "($help)*--dns=-[Set custom dns servers]:dns server: "
> > -        "($help)*--dns-search=-[Set custom DNS search domains]:dns domains: "
> > +        "($help)*--dns=-[Set custom DNS servers]:DNS server: "
> > +        "($help)*--dns-opt=-[Set custom DNS options]:DNS option: " \
>
> Typo, need to remove the \ (this section of code is a array).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16031/files#r39251701>.
>

mrjana commented:
LGTM

thaJeztah commented on docs/articles/networking.md:
```diff
@@ -215,12 +218,17 @@ Four different options affect container domain name services.
+ *  `--dns-opt=OPTION...` — sets the options used by DNS resolvers
+    by writing an `options` line into the container's `/etc/resolv.conf`.
+    See documentation for `resolv.conf` for a list of valid options.  Use
+    `--dns-opt=''` if you don't wish to set options.
```
If I *don't* add a `--dns-opt`, will there still be an `options` line written? I.e., is *omitting* `--dns-opt` different than adding `--dns-opt=''`? If there's no difference, then I'd drop the last line.

thockin replied:
good point.  I got too loose with cut-paste from dns-search

thaJeztah commented on docs/reference/api/docker_remote_api_v1.15.md:
```diff
@@ -562,6 +564,7 @@ Json Parameters:
       a boolean value.
 -   **Dns** - A list of dns servers for the container to use.
 -   **DnsSearch** - A list of DNS search domains
+-   **DnsOptions** - A list of DNS options
```
I don't think this PR should affect API versions < v1.21, if it does, should the code be updated to perform a version-check before accepting/outputting this?

thockin replied:
no, you're probably right that it shouldn't.  I am not very familiar with the internal factoring and I was not sure if it would show up as an additive field or not.

thaJeztah commented on docs/reference/api/docker_remote_api_v1.20.md:
```diff
@@ -268,6 +269,7 @@ Json Parameters:
           Specified as a boolean value.
     -   **Dns** - A list of DNS servers for the container to use.
     -   **DnsSearch** - A list of DNS search domains
+    -   **DnsOptions** - A list of DNS options
```
Changes in API version v1.20 should now be moved to v1.21 (which is the version to be used in Docker 1.9)

thockin replied:
got it, next push

thaJeztah commented on man/docker-create.1.md:
```diff
@@ -117,6 +118,9 @@ two memory nodes.
    Set custom DNS servers
 
+**--dns-opt**=[]
+   Set custom DNS options (Use --dns-opt='' if you don't wish to set options)
```
Same question here; is `--dns-opt=''` different from omitting `--dns-opt`?

thockin replied:
fixed

thaJeztah commented on man/docker-run.1.md:
```diff
@@ -184,6 +185,9 @@ stopping the process by pressing the keys CTRL-P CTRL-Q.
    Set custom DNS search domains (Use --dns-search=. if you don't wish to set the search domain)
 
+**--dns-opt**=[]
+   Set custom DNS options (Use --dns-opt='' if you don't wish to set options)
```
Same here

thaJeztah commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -940,7 +940,7 @@ func (s *DockerSuite) TestRunDnsDefaultOptions(c *check.C) {
 func (s *DockerSuite) TestRunDnsOptions(c *check.C) {
 	testRequires(c, DaemonIsLinux)
-	out, stderr, _ := dockerCmdWithStdoutStderr(c, "run", "--dns=127.0.0.1", "--dns-search=mydomain", "busybox", "cat", "/etc/resolv.conf")
+	out, stderr, _ := dockerCmdWithStdoutStderr(c, "run", "--dns=127.0.0.1", "--dns-search=mydomain", "--dns-opt=ndots:9", "busybox", "cat", "/etc/resolv.conf")
```
Just looking at this; do we need a test;

- using *multiple* `--dns-opt` options?
- a test setting these options on the *daemon* (e.g., similar to https://github.com/docker/docker/blob/master/integration-cli/docker_cli_daemon_test.go#L1146), and overriding them on a container?

thockin replied:
I'm adding a test for repeated options.

As for daemon flags, I tried to follow suit with the existing DNS tests.

thaJeztah commented on daemon/daemon_test.go:
```diff
@@ -458,7 +458,7 @@ func TestRemoveLocalVolumesFollowingSymlinks(t *testing.T) {
 
 	hostConfig := `{"Binds":[],"ContainerIDFile":"","LxcConf":[],"Memory":0,"MemorySwap":0,"CpuShares":0,"CpusetCpus":"",
-"Privileged":false,"PortBindings":{},"Links":null,"PublishAllPorts":false,"Dns":null,"DnsSearch":null,"ExtraHosts":null,"VolumesFrom":null,
+"Privileged":false,"PortBindings":{},"Links":null,"PublishAllPorts":false,"Dns":null,"DnsOptions":null,"DnsSearch":null,"ExtraHosts":null,"VolumesFrom":null,
```
Sorry for being a PITA, and I'm not a code reviewer; but I don't see any tests being added to test the new option, only making the existing tests pass; do we need tests?

thockin replied:
This test doesn't test any permutations of DNS

thaJeztah replied:
Yeah, I just picked a random line in this file to add my comment. I'm wasnt sure if tests were needed, just noticed that (afaics) there were no tests to check if `--dns-opt` at the daemon were successfully applied (and could be overridden from a container).

thaJeztah commented:
Thanks @thockin, I added some comments inline, in addition;

- the [`man/docker.1.md`](https://github.com/docker/docker/blob/master/man/docker.1.md) file needs to be updated for setting the option at startup. Note that this file is being worked on, and these options will be moved to a separate man page (see https://github.com/docker/docker/pull/16063)
- the [docs/reference/commandline/create.md](https://github.com/docker/docker/blob/master/docs/reference/commandline/create.md),
- [docs/reference/commandline/run.md](https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md),
- and [docs/reference/commandline/daemon.md](https://github.com/docker/docker/blob/master/docs/reference/commandline/daemon.md) need the new option mentioned.
- having an actual example would be nice (if it makes sense :-))

I added some questions wrt changing the behavior of older API versions; https://github.com/docker/docker/pull/16031/files#discussion_r39347362, and tests (https://github.com/docker/docker/pull/16031/files#discussion_r39347533 and https://github.com/docker/docker/pull/16031/files#discussion_r39347642). I'm no code reviewer, but perhaps @cpuguy83 or @calavera have an opinion on this :-)

calavera commented:
I'm okay with having those references in old inspect structs. Currently, we don't have a dedicated struct for inspecting host configurations, and that should not be in the scope of this PR.

thockin commented:
Woops, I stripped them out before reading all the comments.  Decision?  I
can put them back...

On Mon, Sep 14, 2015 at 9:41 AM, David Calavera <notifications@github.com>
wrote:

> I'm okay with having those references in old inspect structs. Currently,
> we don't have a dedicated struct for inspecting host configurations, and
> that should not be in the scope of this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16031#issuecomment-140139420>.
>

calavera commented:
@thockin it's fine, I'm pretty sure we have other fields propagated to old api versions that are not fully documented. I think the important one is the version that introduces the field.

thockin commented:
I think all comments are addressed - please shout if I missed anything.  New push is up.

thaJeztah commented:
Thanks @thockin, I think these are still needed;

- [docs/reference/commandline/create.md](https://github.com/docker/docker/blob/master/docs/reference/commandline/create.md),
- [docs/reference/commandline/run.md](https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md),
- [docs/reference/commandline/daemon.md](https://github.com/docker/docker/blob/master/docs/reference/commandline/daemon.md)
- a mention in the [API changelog](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v121-api-changes)

Sorry for this being such an extensive list, thanks in advance :heart:

thockin commented:
Not sure how I missed those in my search-replace.  Thanks.  Done.

thaJeztah commented:
Thanks so much @thockin, LGTM!

ping @moxiegirl @SvenDowideit

moxiegirl commented:
LGTM

calavera commented:
Please, do not merge until we get a green build. I'd like to make sure tests still pass. :pray: :space_invader:

runcom commented:
sorry this needs a rebase :( :( :( :*(

thockin commented:
Done.  It was just the changelog

runcom commented:
cool @thockin merging!

runcom commented:
@calavera tests green, merging :tada:

thaJeztah commented:
Great, thanks @thockin !
